Ruby3

LocoDelAssembly opened the pull request:
- Increased target to Ruby 2.5 as required by Rubocop
- Enabled Ruby 3.x compatibility
- Enabled CI at Github Actions for Ruby versions 2.5, 2.6, 2.7 and 3.0
- Some code refactoring to make Rubocop happy

LocoDelAssembly commented:
For tests results you may check https://github.com/LocoDelAssembly/gpx/actions (logs are retained for a maximum of 90 days). Github Actions needs to be enabled on this repo to enable testing on pull requests and pushes to `master`.

moveson commented:
It would seem that for this gem to continue to be viable, it needs to make the jump to Ruby 3. If the maintainers intend to merge this well-tested PR, would someone please have a look? Or perhaps additional maintainers could be brought in?

moveson commented on lib/gpx/bounds.rb:
```diff
@@ -7,6 +7,7 @@ class Bounds < Base
     # Creates a new bounds object with the passed-in min and max longitudes
     # and latitudes.
     def initialize(opts = { min_lat: 90.0, max_lat: -90.0, min_lon: 180.0, max_lon: -180.0 })
+      super()
```
Idiomatic Ruby would be to use `super` without the `()`

LocoDelAssembly replied:
```ruby
diff --git a/lib/gpx/bounds.rb b/lib/gpx/bounds.rb
index f6bb81c..f0c2761 100644
--- a/lib/gpx/bounds.rb
+++ b/lib/gpx/bounds.rb
@@ -7,7 +7,7 @@ module GPX
     # Creates a new bounds object with the passed-in min and max longitudes
     # and latitudes.
     def initialize(opts = { min_lat: 90.0, max_lat: -90.0, min_lon: 180.0, max_lon: -180.0 })
-      super()
+      super
       @min_lat = opts[:min_lat].to_f
       @max_lat = opts[:max_lat].to_f
       @min_lon = opts[:min_lon].to_f
```

```
# Running:

EEEEEE.EEEEEEEEEEE...EEEEEEEEE....EEEEE

Finished in 0.034327s, 1136.1240 runs/s, 320.4452 assertions/s.

  1) Error:
GPXFileTest#test_big_file:
ArgumentError: wrong number of arguments (given 1, expected 0)
    /home/hernan/Projects/LocoDelAssembly/gpx/lib/gpx/bounds.rb:10:in `initialize'
    /home/hernan/Projects/LocoDelAssembly/gpx/lib/gpx/bounds.rb:10:in `initialize'
    /home/hernan/Projects/LocoDelAssembly/gpx/lib/gpx/gpx_file.rb:190:in `new'
    /home/hernan/Projects/LocoDelAssembly/gpx/lib/gpx/gpx_file.rb:190:in `reset_meta_data'
    /home/hernan/Projects/LocoDelAssembly/gpx/lib/gpx/gpx_file.rb:48:in `initialize'
    /home/hernan/Projects/LocoDelAssembly/gpx/tests/gpx_file_test.rb:50:in `new'
    /home/hernan/Projects/LocoDelAssembly/gpx/tests/gpx_file_test.rb:50:in `test_big_file'
```

:)

andrewhao commented:
Hi @LocoDelAssembly - thank you for your contribution. I'm in agreement with your changes and keeping up with Ruby versions.

I have a minor comment to make here, and that with this PR we will now be split between Travis CI and Github.

I'm personally not as familiar with the Github CI workflow, but it seems like we're following a pretty documented path [here](https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-ruby), so I trust you on this. If we do choose to merge this, the project will need to take the step of deprecating TravisCI in favor of Github. Given that the project is in the process of seeking new maintainers (#44), my preference is not to rock the boat too much in terms of our workflow.

Here's my preference - I'd prefer you try to make this build work with Travis to reduce infrastructure complexity in this PR. Feel free to rip out the tests for Ruby 2.2 and any other unsupported versions in the existing tests (it seems like the build hasn't been working recently?)

If this is something that you feel is difficult, then I'm amenable to going with Github, then asking the new maintenance team to make the switch. What do you think?

cc @niborg

LocoDelAssembly commented:
Hi @andrewhao. I have updated Travis CI build to use from Ruby 2.5 to 3.0 (commented-out 3.1 as it is not supported yet there).

Seems that Travis CI is still "charging" me credits like before when I stop using it on all my projects and moved to GitHub Actions which is free for all public repos. If you can get Travis CI to be truly free then I guess either CI is fine, otherwise I strongly suggest moving to GitHub Actions.

Ruby 2.5 might need to be dropped as it is likely in EOL now.

# Latest runs
GitHub: https://github.com/LocoDelAssembly/gpx/actions/runs/1633855691
Travis: https://app.travis-ci.com/github/LocoDelAssembly/gpx/builds/244083947

niborg commented:
> If we do choose to merge this, the project will need to take the step of deprecating TravisCI in favor of Github.
I can look into this. It does look like Github Actions provides a lot of conveniences since I last looked at it, free CI being one of them.

andrewhao commented:
Thank you @LocoDelAssembly - Travis is free for OSS, but it certainly is nice having CI and code all in one place. Thank you!

moveson commented:
Do you plan to cut a new release?
